neuroml/nml/helper_methods.py

```diff
@@ -2294,8 +2298,265 @@ def get_segment_adjacency_list(self):
             except AttributeError:
                 print(f"Warning: Segment: {segment} has no parent")
 
+        self.adjacency_list = child_lists
         return child_lists
 
+    def get_graph(self):
+        """Get a networkx DiGraph of the morphology of the cell with distances
+        between the proximal point of a parent and the point where a child
+        connects to it as the weights of the edges of the graph.
+
+        Please see https://networkx.org/documentation/stable/reference
+        for information on networkx routines that can be used on this graph.
+
+        This method also stores the graph in the `self.cell_graph` attribute
+        for future use.
+
+        :returns: networkx.Graph
+
+        """
+        import networkx as nx
+        import math
+        cell_graph = nx.DiGraph()
+
+        # don't recompute if already exists
+        adlist = getattr(self, "adjacency_list", None)
+        if adlist is None:
+            adlist = self.get_segment_adjacency_list()
+
+        for parid, childrenids in adlist.items():
+
+            par_length = self.get_segment_length(parid)
+
+            for cid in childrenids:
```

Comment:
I wonder how inefficient this type of thing could be for complex morphos - there are nested loops in different places here. If some of these were converted to numpy arrays, we could get rid of most of the loops? networkx also likes edge lists or just an adj matrix. But maybe all that is just premature optimisation...

Reply:
I do use an adjacency matrix here (which we'd implemented as part of `sectionise` etc. previously) and that doesn't seem to be very expensive---it does one pass over all segments:

https://github.com/NeuralEnsemble/libNeuroML/blob/development/neuroml/nml/helper_methods.py#L2274

No way around calculation of all segment lengths, I don't think :thinking:

Reply:
Sorry, let me try to be more specific (seeing this code for the first time in many years...)  Some of these utility methods rely on similar (and potentially expensive) operations over the morphology - eg all operations involving the conn graph must build the graph itself on the first place. There will be user workflows in which more than one of these methods will be invoked, leading to duplicated code/computations. I was wondering if it would make sense to have 'stateful' methods - which is precisely what you are doing for the adjacency list, and I thought it would make sense for some of these properties to be numpy arrays - as we will be doing broadcastable operations eg on things involving geometry.

Reply:
I didn't see that you had already relied on things like self.adjacency_list - so my bad for only reading the diffs for this PR instead of the whole thing!

Reply:
Yeh, I try to cache info where possible---did that for the adjacency list and the graph bits. 

I haven't looked into making the object model more numpy friendly so we can take advantage of numpy's broadcasting etc. I think it'll need some investigation---to see how it'll all work with generateds and so on?